chain/chain/src/store.rs

```diff
@@ -440,25 +440,70 @@ impl ChainStore {
         )
     }
 
+    // Get outgoing receipts that will be sent from shard `shard_id` from block whose prev block
+    // is `prev_block_hash`
+    // Note that the meaning of outgoing receipts here are slightly different from save_outgoing_receipts
+    // There, outgoing receipts are generated after they are generated from a chunk from the last block
+    // at height h-1
+    // Here, outgoing receipt are about to be send from shard with `shard_id` for block at height h
+    // The difference is important because shard layout may change between epochs, or, between
+    // when outgoing_receipts are saved and when this function is called
+    // Note, the current way of implementation only works when there are two shard layouts, or
+    // at least one chunk is generated before shard layout changed again when shard layout changed
+    // more than twice. This is not a problem right now for simple nightshade migration, since
+    // we are changing shard layout for the first time.
+    // But we need to implement a more general algorithm should we upgrade shard layout again
+    // in the future
     pub fn get_outgoing_receipts_for_shard(
         &mut self,
+        runtime_adapter: &dyn RuntimeAdapter,
         prev_block_hash: CryptoHash,
         shard_id: ShardId,
         last_included_height: BlockHeight,
-    ) -> Result<ReceiptResponse, Error> {
+    ) -> Result<Vec<Receipt>, Error> {
+        let shard_layout = runtime_adapter.get_shard_layout_from_prev_block(&prev_block_hash)?;
         let mut receipts_block_hash = prev_block_hash;
         loop {
             let block_header = self.get_block_header(&receipts_block_hash)?;
 
             if block_header.height() == last_included_height {
+                let receipts_shard_layout =
+                    runtime_adapter.get_shard_layout(block_header.epoch_id())?;
+
+                // get the shard from which the outgoing receipt were generated
+                let receipts_shard_id = if shard_layout != receipts_shard_layout {
+                    // we unwrap here because otherwise it means the shard layout is invalid
+                    shard_layout.get_parent_shard_id(shard_id).unwrap()
+                } else {
+                    shard_id
+                };
                 let receipts = if let Ok(cur_receipts) =
```

Comment:
http://sled.rs/errors.html

```rust
enum Error {
  Local(LocalError),
  Fatal(FatalError),
}
```

it feels exactly like what we are doing here :D

Reply:
Yeah I also found it weird. I think the error should be propagated through. I don't want to change it in this PR though since I want to preserve the behavior. I will change it in another PR.

Reply:
Yeah personally I have always struggled to find a good way to understand how errors should be handled. Maybe this deserves a larger scope discussion so that everyone working on the code is on the same page for error handling. @bowenwang1996